File: binarytree.c
#include <stdio.h>

struct node {
	struct node *left;
	struct node *right;
	int value;
};

struct node *stack[200];
int _top = 0;
int empty() {
	return _top == 0;
}
void push(struct node *p) {
	stack[_top] = p;
	_top++;
}
struct node *pop() {
	--_top;
	return stack[_top];
}
struct node *top() {
	return stack[_top-1];
}
void in_order(struct node *root) {
	struct node *p = root;
	while(p || !empty()) {
		if(p) {
			push(p);
			p = p->left;
		} else {
			p = pop();
			printf("%d ",p->value);
			p = p->right;
		}
	}
}
void in_order_cur(struct node *root) {
	if(root == NULL)
		return;
	in_order_cur(root->left);
	printf("%d ",root->value);
	in_order_cur(root->right);
}
/* 只有从右树回来的时候才可以访问，第一次遇到的时候不能访问
 * 记一个lastvist用于做这个事情
 */
void post_order(struct node *root) {
	struct node *lastvisit, *p;

	lastvisit = NULL;
	p = root;
	while(p || !empty()) {
		if(p) {
			push(p);
			p = p->left;
		} else {
			p = top();
			if(!p->right || p->right==lastvisit) {
				p = pop();
				printf("%d ",p->value);
				lastvisit = p;
				p = NULL;
			} else {
				p = p->right;
			}
		}
	}
}
int main() {
	struct node root,a,b,c,d,e,f;

	root.value = 1;
	a.value = 2;
	b.value = 3;
	c.value = 4;
	d.value = 5;
	e.value = 6;
	f.value = 7;

	root.left = &a;
	root.right = &b;
	a.left = &c;
	a.right = &d;
	c.left = c.right = NULL;
	d.left = NULL;
	d.right = &e;
	e.left = e.right = NULL;
	b.left = NULL;
	b.right = &f;
	f.left = f.right = NULL;
	
	post_order(&root);
	return 0;
}

File: peral/binarytree.go
package main

import (
	"fmt"
)

type Node struct {
	V     byte
	Left  *Node
	Right *Node
}

type Stack struct {
	arr []*Node
	top int
}

func NewStack(cap int) *Stack {
	ret := &Stack{
		arr: make([]*Node, cap),
	}
	return ret
}

func (s *Stack) Top() *Node {
	return s.arr[s.top-1]
}

func (s *Stack) Push(n *Node) {
	s.arr[s.top] = n
	s.top++
}

func (s *Stack) Pop() *Node {
	s.top--
	return s.arr[s.top]
}

func (s *Stack) Empty() bool {
	return s.top <= 0
}

func PreOrderNaive(tree *Node) {
	if tree == nil {
		return
	}

	fmt.Printf("%c", tree.V)
	PreOrderNaive(tree.Left)
	PreOrderNaive(tree.Right)
}

func PreOrder(tree *Node) {
	if tree == nil {
		return
	}

	s := NewStack(20)
	s.Push(tree)

	for !s.Empty() {
		n := s.Pop()
		fmt.Printf("%c", n.V)

		if n.Right != nil {
			s.Push(n.Right)
		}

		if n.Left != nil {
			s.Push(n.Left)
		}
	}
}

func MidOrder(root *Node) {
	s := NewStack(20)

	for ; root != nil; root = root.Left {
		s.Push(root)
	}

	for !s.Empty() {
		n := s.Pop()
		fmt.Printf("%c", n.V)

		for n = n.Right; n != nil; n = n.Left {
			s.Push(n)
		}
	}
}

type Queue struct {
	arr []*Node
	tail int
	front int
}

func NewQueue(cap int) *Queue {
	return &Queue{
		arr: make([]*Node, cap),
	}
}

func (q *Queue) Empty() bool {
	return q.front == q.tail
}

func (q *Queue) Push(n *Node) {
	q.arr[q.tail] = n
	q.tail = (q.tail + 1) % len(q.arr)
	return
}

func (q *Queue) Pop() *Node {
	ret := q.arr[q.front]
	q.front = (q.front + 1) % len(q.arr)
	return ret
}

func LevelOrder(root *Node) {
	q := NewQueue(20)
	q.Push(root)
	last := root
	
	for !q.Empty() {
		n := q.Pop()
		fmt.Printf("%c", n.V)
		
		if last == n {
			fmt.Println()
			if n.Right != nil {
				last = n.Right
			} else {
				last = n.Left
			}
		}
		
		if n.Left != nil {
			q.Push(n.Left)
		}
		if n.Right != nil {
			q.Push(n.Right)
		}
	}
}

func PostOrder(root *Node) {
	p := root
	var lastVisit *Node
	s := NewStack(20)

	for p != nil || !s.Empty() {
		if p != nil {
			s.Push(p)
			p = p.Left
		} else {
			p = s.Top()
			if p.Right == nil || p.Right == lastVisit {
				s.Pop()
				fmt.Printf("%c", p.V)
				lastVisit = p
				p = nil
				continue
			}
			p = p.Right
		}
	}
}

func main() {
	a := &Node{V: 'a'}
	b := &Node{V: 'b'}
	c := &Node{V: 'c'}
	d := &Node{V: 'd'}
	e := &Node{V: 'e'}
	f := &Node{V: 'f'}
	g := &Node{V: 'g'}

	a.Left = b
	a.Right = c
	b.Left = d
	b.Right = e
	e.Right = f
	c.Left = g

	LevelOrder(a)
}
